### PR TITLE
Add application submission id everywhere, with an optional prefix

### LINKED ISSUE #3830: Allow adding prefix to IDs of submissions per fund/lab
**Is your feature request related to a problem? Please describe.**

No.

**Is your feature request related to an existing functionality? Please describe.**

The IDs of the submissions is auto-incremented number right now. This is not ideal for referencing an application outside the hypha software. Having an optional prefix/shortcode, like `FUND24-<num>` allows for an easily identifiable identifier for application which can be referenced outside the hypha system without confusion.

**Describe the solution you'd like**

- In admin, an ID prefix can be added
- After the prefix is set, then all the submissions made to that round/lab will store a `public_id` constructed using the prefix and the original db id. 
- The `public_id` will be searchable and displayed everywhere instead of submission -> id.

**Describe alternatives you've considered**

Generate the submission id dynamically instead of storing it. This is not ideal, as it can be modified very easily, and won't lead to a permanent reference number. 

**Affected roles**

- Staff
- Applicants
- Reviewers
- Partners
- Developers
- General public

**Ideal deadline**

This Month

### PR DESCRIPTION
**Pull request by theskumar**
This PR allow adding an option prefix to the submission ID's that are auto generated. 

The prefix can be added by going to round or lab settings page. 

This PR also ensure the prefix or application ID displayed along with title almost everywhere for each reference. Since the title of application is not unique, this helps add uniqueness to it. 

The autolinking of submission ids in communication has be updated to account for application id prefix

Fixes https://github.com/HyphaApp/hypha/issues/3830

Screenshot 2024-05-20 at 3  38 36@2x

Screenshot 2024-05-20 at 4  43 55@2x

Screenshot 2024-05-20 at 4  44 32@2x
Screenshot 2024-05-20 at 4  45 05@2x

## Test Steps

 - [ ] general checks across the site
 - [ ] check notification generate - slack, email, etc.

### CODE REVIEW

File: hypha/apply/funds/models/submissions.py [hunk 1]
```diff
@@ -507,6 +510,10 @@ class Meta:
     def is_draft(self):
         return self.status == DRAFT_STATE
 
+    @property
+    def title_with_id(self):
```
Comment (sandeepsajan0) [inline hypha/apply/funds/models/submissions.py]: Can we use this method to provide 'title' or 'title with public_id' based on a feature flag? Some org might not want to show the application count or not at least to the applicant.

Comment (theskumar) [reply hypha/apply/funds/models/submissions.py]: Great suggestion, however, if possible I would try keep the number of feature flags to minimal, unless any adaptor thinks it's a deal break for them. 

@wes-otf @frjo If you happen to present this in adopter meeting, can we try get an idea if someone would want to hide this info? The ids have been and will always be available in the url.

File: hypha/apply/funds/templates/funds/includes/rendered_answers.html [hunk 2]
```diff
@@ -25,13 +25,13 @@ <h5 class="text-base">{% trans "E-mail" %}</h5>
...
     {% if object.get_address_display != "-" %}
-        <div class="grid__cell--span-two">
+        <div>
             <h5 class="text-base">{% trans "Address" %}</h5>
             {{ object.get_address_display }}
         </div>
     {% endif %}
     {% if object.get_organization_name_display != "-" %}
-        <div class="grid__cell--span-two">
+        <div class="">
```
Comment (frjo) [inline hypha/apply/funds/templates/funds/includes/rendered_answers.html]: Remove the class attribute.

File: hypha/apply/funds/models/submissions.py (hunk 1 above)
Comment (frjo) [reply hypha/apply/funds/models/submissions.py]: Did not sound like a need to hide exist. Could we provide a template for this? Making it easy to adopt the output.

You mentioned that some wanted to have the prefix first.

**Review by frjo — CHANGES_REQUESTED** [review]
Do we support search for prefix or number (\#123)?

File: hypha/apply/funds/models/submissions.py (hunk 1 above)
Comment (theskumar) [reply hypha/apply/funds/models/submissions.py]: Updated in https://github.com/HyphaApp/hypha/pull/3930/commits/0e406525490d9609c20cf2d0850f23314561adcb

**Comment by theskumar** [comment]
Getting the submission prefix in the url as well is lot of work also requires decent testing. I think it would better to tackled in another pr.

**Comment by frjo** [comment]
@theskumar I think this needs rebuild migrations and a rebase. Merged in a number of PRs today.

**Comment by theskumar** [comment]
@frjo updated! :)

**Review by frjo — APPROVED** [review]

File: hypha/apply/activity/adapters/slack.py [hunk 3]
```diff
@@ -31,51 +31,55 @@ class SlackAdapter(AdapterBase):
...
-            "A new submission has been submitted for {source.page.title}: <{link}|{source.title}> by {user}"
+            "A new submission has been submitted for {source.page.title}: <{link}|{source.title_text_display}> by {user}"
         ),
         MESSAGES.UPDATE_LEAD: _(
-            "The lead of <{link}|{source.title}> has been updated from {old_lead} to {source.lead} by {user}"
+            "The lead of <{link}|{source.title_text_display}> has been updated from {old_lead} to {source.lead} by {user}"
         ),
         MESSAGES.BATCH_UPDATE_LEAD: "handle_batch_lead",
         MESSAGES.COMMENT: _(
             "A new {comment.visibility} comment has been posted on <{link}|{source.title}> by {user}"
```
Comment (wes-otf) [inline hypha/apply/activity/adapters/slack.py:41]: Should this be:
```suggestion
            "A new {comment.visibility} comment has been posted on <{link}|{source.title_text_display}> by {user}"
```
That's only one I noticed that was missing on slack

**Comment by wes-otf** [comment]
Super small issue from my testing with the slack comment notif but otherwise works great!

**Pull request merged**